external-crates/move/crates/move-binary-format/src/compatibility.rs

```diff
@@ -208,25 +394,20 @@ impl Compatibility {
         // friend list. But for simplicity, we decided to go to the more restrictive form now and
         // we may revisit this in the future.
         for (name, old_func) in &old_module.functions {
+            // if we don't find a function in the new module
             let Some(new_func) = new_module.functions.get(name) else {
-                if old_func.visibility == Visibility::Friend {
-                    friend_linking = false;
-                } else if old_func.visibility != Visibility::Private {
-                    datatype_and_function_linking = false;
-                } else if old_func.is_entry && self.check_private_entry_linking {
-                    // This must be a private entry function. So set the link breakage if we're
-                    // checking for that.
-                    entry_linking = false;
-                }
+                context.function_missing(old_func, self.check_private_entry_linking);
                 continue;
             };
 
             // Check visibility compatibility
             match (old_func.visibility, new_func.visibility) {
                 (Visibility::Public, Visibility::Private | Visibility::Friend) => {
-                    datatype_and_function_linking = false
+                    context.function_visibility_mismatch(old_func, new_func);
+                }
+                (Visibility::Friend, Visibility::Private) => {
+                    context.function_friend_linking(old_func);
```

Comment:
I think you'll need a new function here as well -- `function_lost_friend_visibility`?